File: Interview/DS/Sorting/Insertion_Sort/Insertion_sort.java
import java.util.Arrays;

public class InsertionSort {
    public static void main(String[] args) {
        int[] arr = {5, 1, 4, 2, 8};
        insertionSort(arr);
        System.out.println("Sorted array: " + Arrays.toString(arr));
    }

    /**
     * The main insertion sort method that takes an array as its argument.
     * 
     * @param arr the array to be sorted
     */
    public static void insertionSort(int[] arr) {
        int n = arr.length;
        for (int i = 1; i < n; i++) {
            int key = arr[i];
            int j = i - 1;
            
            // move elements of arr[0..i-1], that are greater than key, to one position ahead of their current position
            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j = j - 1;
            }
            arr[j + 1] = key;
        }
    }
}

File: Interview/Algorithms/Sorting/Insertion_Sort/insertion_sort.py
def insertionSort(arr):
    # loop through the array, starting from the second element
    for i in range(1, len(arr)):
        current_value = arr[i] # store the current value
        position = i # store the current position
        # move the elements of the sorted part of the array that are larger than the current value to the right
        while position > 0 and arr[position - 1] > current_value:
            arr[position] = arr[position - 1]
            position -= 1
        # insert the current value in the correct position
        arr[position] = current_value
    return arr

# Example usage
arr = [64, 34, 25, 12, 22, 11, 90]
print("Sorted array:", insertionSort(arr))
